UnitTestBot/UTBotJava
# Improve detection of NPEs caused by this instance fields being null in Spring unit tests

## Description

Fixes #2589 (desired NPE test is generated by fuzzer, because symbolic engine has troubles dealing with `getClass() == o.getClass()` condition.

To make fuzzer able to generate `Object` of type `Order` special `JavaLangObjectValueProvider` was added that tries to use `classUnderTest` whenever `java.lang.Object` is needed (plain `new Object()` can still be used where `java.lang.Object` is needed).

In case class under test has some hard to pass validation inside of the constructor and fuzzer is asked to create multiple `java.lang.Object`s, [`RemovingConstructFailsUtExecutionInstrumentation`](https://github.com/UnitTestBot/UTBotJava/commit/c74bf37a9330b1ba0c4127724c61608705f8b114) was used in all Spring tests (in short, it detects impossible to create due to exceptions thrown by constructor models in `stateBefore` and replaces them with `UtNullModel`s, making it easier for fuzzer to find "creatable" input and actually call method under test).

Finally, `InjectMocksValueProvider` was made to respect `NonNullSpeculator` (previously fuzzer considered all fields to be non-`null`).

## How to test

### Manual tests

NPE unit be generated by fuzzer for `Order.equals()` from `spring-boot-testing` project.

Generate unit tests for the following classes individually with fuzzer only and with engine only, each of them should generate NPE tests for methods marked with "NPE expected" and no NPE tests for methods marked with "no NPE expected".

```java
package com.rest.order.services;

import com.rest.order.models.Order;
import com.rest.order.repositories.OrderRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.lang.Nullable;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ServiceExample {
    /// region notAutowired

    private OrderRepository notAutowired;

    public void setNotAutowired(OrderRepository notAutowired) {
        this.notAutowired = notAutowired;
    }

    @Autowired
    public void fakeSetNotAutowired(OrderRepository notAutowired) {}

    @Autowired
    public void setNonRequiredNotAutowired(@Autowired(required = false) OrderRepository notAutowired) {
        this.notAutowired = notAutowired;
    }

    @Autowired(required = false)
    public void nonRequiredSetNotAutowired(OrderRepository notAutowired) {
        this.notAutowired = notAutowired;
    }

    @Autowired
    public void setNullable(@Nullable OrderRepository notAutowired) {
        this.notAutowired = notAutowired;
    }

    public ServiceExample(OrderRepository notAutowired) {
        this.notAutowired = notAutowired;
    }

    @Autowired(required = false)
    public ServiceExample(OrderRepository notAutowired, boolean unused) {
        this.notAutowired = notAutowired;
    }

    @Autowired
    public ServiceExample(@Autowired(required = false) OrderRepository notAutowired, char unused) {
        this.notAutowired = notAutowired;
    }

    // NPE expected
    public List<Order> useNotAutowired() {
        return notAutowired.findAll();
    }

    /// endregion

    /// region autowiredViaConstructor

    private OrderRepository autowiredViaConstructor;

    @Autowired
    public ServiceExample(OrderRepository orderRepository, int unused) {
        this.autowiredViaConstructor = orderRepository;
    }

    // no NPE expected
    public List<Order> useAutowiredViaConstructor() {
        return autowiredViaConstructor.findAll();
    }

    /// endregion

    /// endregion

    /// region autowiredViaField

    @Autowired
    private OrderRepository autowiredViaField;

    // no NPE expected
    public List<Order> useAutowiredViaField() {
        return autowiredViaField.findAll();
    }

    /// endregion

    /// region autowiredViaSetter

    private OrderRepository autowiredViaSetter;

    @Autowired
    public void setAutowiredViaSetter(OrderRepository orderRepository) {
        this.autowiredViaSetter = orderRepository;
    }

    // no NPE expected
    public List<Order> useAutowiredViaSetter() {
        return autowiredViaSetter.findAll();
    }

    /// endregion

}
```

```java
package com.rest.order.services;

import com.rest.order.models.Order;
import com.rest.order.repositories.OrderRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ServiceExample2 {
    private final OrderRepository autowiredViaOnlyConstructor;

    public ServiceExample2(OrderRepository orderRepository) {
        this.autowiredViaOnlyConstructor = orderRepository;
    }

    // no NPE expected
    public List<Order> useAutowiredViaOnlyConstructor() {
        return autowiredViaOnlyConstructor.findAll();
    }
}
```

```java
package com.rest.order.services;

import com.rest.order.models.Order;
import com.rest.order.repositories.OrderRepository;

import java.util.List;

public class ServiceExample3 {
    private final OrderRepository notAutowiredViaOnlyConstructor;

    public ServiceExample3(OrderRepository orderRepository) {
        this.notAutowiredViaOnlyConstructor = orderRepository;
    }

    // NPE expected
    public List<Order> useNotAutowiredViaOnlyConstructor() {
        return notAutowiredViaOnlyConstructor.findAll();
    }
}
```

```java
package com.rest.order.services;

import com.rest.order.models.Order;
import com.rest.order.repositories.OrderRepository;
import org.springframework.lang.Nullable;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ServiceExample4 {
    public OrderRepository notAutowiredViaOnlyConstructor;
    private final OrderService orderService;

    public ServiceExample4(@Nullable OrderRepository orderRepository, OrderService orderService) {
        this.notAutowiredViaOnlyConstructor = orderRepository;
        this.orderService = orderService;
    }

    // NPE expected
    public List<Order> useNotAutowiredViaOnlyConstructor() {
        return notAutowiredViaOnlyConstructor.findAll();
    }
}
```

## Self-check list
